04-Protocol-Security.md

```diff
@@ -0,0 +1,407 @@
+4. compute `P` as `d'⋅G`
+5. drop the Y coordinate and output keypair `(sk, bytes(P.x))`
+
+Such system has the following properties:
+* for each keypair `(sk, bytes(P.x))` there is another keypair `(n - sk, bytes(P.x))`, where `n` is group order of secp256k1 curve
+* Each result of `ECDH(sk, Q)` is equal to `ECDH(n - sk, Q)` for some EC point `Q` where `n` is group order of secp256k1 curve
+
+These properties don't reduce security.
+
+For more information refer to BIP340<sup>[3](#reference-3)</sup>
+
+### 4.3.2 Hash function
+* `SHA-256()` is used as a `HASH()`
+
+### 4.3.3 Cipher function for authenticated encryption
+* Cipher has methods for encryption and decryption for key `k`, nonce `n`, associated_data `ad`, plaintext `pt` and ciphertext `ct`
+  * `ENCRYPT(k, n, ad, pt)`
+  * `DECRYPT(k, n, ad, ct)`
+* ChaCha20 and Poly1305 in AEAD mode<sup>[4](reference-4)</sup> (ChaChaPoly) is used as a default AEAD cipher
+
+## 4.4 Cryptographic operations
+
+### 4.4.1 CipherState object
+Object that encapsulates encryption and decryption operations with underlying AEAD mode cipher functions using 32-byte encryption key `k` and 8-byte nonce `n`.
+CipherState has the following interface:
+
+* `InitializeKey(key)`:
+  * Sets `k = key`, `n = 0`
+* `EncryptWithAd(ad, plaintext)`
+  * if `k` is non-empty, performs `ENCRYPT(k, n++, ad, plaintext)` on the underlying cipher function, otherwise returns `plaintext`
+* `DecryptWithAd(ad, ciphertext)`
+  * if `k` is non-empty performs `DECRYPT(k, n++, ad, plaintext)` on the underlying cipher function, otherwise returns ciphertext. If an authentication failure occurs in `DECRYPT()` then `n` is not incremented and an error is signaled to the caller.
+
+### 4.4.2 Handshake Operation
+Throughout the handshake process, each side maintains these variables:
+
+* `ck`: **chaining key**. Accumulated hash of all previous ECDH outputs. At the end of the handshake `ck` is used to derive encryption key `k`.
+*  `h`: **handshake hash**. Accumulated hash of _all_ handshake data that has been sent and received so far during the handshake process
+* `e`, `re` **ephemeral keys**. Ephemeral key and remote party's ephemeral key, respectively.
+* `s`, `rs` **static keys**. Static key and remote party's static key, respectively.
+
+The following functions will also be referenced:
+
+* `generateKey()`: generates and returns a fresh `secp256k1` keypair
+  * Where the object returned by `generateKey` has two attributes:
+    * `.pub`, which returns an abstract object representing the public key
+    * `.priv`, which represents the private key used to generate the public key
+  * Where the object also has a single method:
+    * `.serializeImplicit()` that outputs a 32-byte serialization of the X-coordinate of EC point (implicit Y-coordinate)
+
+* `a || b` denotes the concatenation of two byte strings `a` and `b`
+
+* `MixKey(input_key_material)`: Executes the following steps:
+  * sets `ck, temp_k = HKDF(ck, input_key_material, 2)`
```

Comment:
103: The output of HKDF is of the form (x, y) or (x, y, z) where x, y and z are 32-bytes integers
(actually, hash outputs). Maybe do you mean 
(ck, temp_k) = HKDF(ck, input_key_material, 2)?
Also, a minor issue. MixKey uses HKDF, which is defined later. Should you move the definition of 
MixKey after the definition of HKDF?